pairing/bls12381/gnark/g1.go

```diff
@@ -0,0 +1,142 @@
+package gnark
+
+import (
+	"crypto/cipher"
+	"fmt"
+	"io"
+	"math/big"
+
+	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	"go.dedis.ch/kyber/v4"
+)
+
+var _ kyber.SubGroupElement = &G1Elt{}
+
+// G1Elt is a wrapper around a G1 point on the BLS12-381 Gnark curve.
+type G1Elt struct{ inner bls12381.G1Jac }
+
+// MarshalBinary returns a compressed point, without any domain separation tag information
+func (p *G1Elt) MarshalBinary() (data []byte, err error) {
+	var g1aff bls12381.G1Affine
+	g1aff.FromJacobian(&p.inner)
+	res := g1aff.Bytes()
+	return res[:], nil
+}
+
+// UnmarshalBinary populates the point from a compressed point representation.
+func (p *G1Elt) UnmarshalBinary(data []byte) error {
+	var g1aff bls12381.G1Affine
+	_, err := g1aff.SetBytes(data)
+	if err != nil {
+		return fmt.Errorf("setting affine representation: %w", err)
+	}
+
+	p.inner.FromAffine(&g1aff)
+	return nil
+}
+
+func (p *G1Elt) String() string { return p.inner.String() }
+
+func (p *G1Elt) MarshalSize() int { return bls12381.SizeOfG1AffineCompressed }
+
+// MarshalTo writes a compressed point to the Writer, without any domain separation tag information
+func (p *G1Elt) MarshalTo(w io.Writer) (int, error) {
+	buf, err := p.MarshalBinary()
+	if err != nil {
+		return 0, err
+	}
+	return w.Write(buf)
+}
+
+// UnmarshalFrom populates the point from a compressed point representation read from the Reader.
+func (p *G1Elt) UnmarshalFrom(r io.Reader) (int, error) {
+	buf := make([]byte, p.MarshalSize())
+	n, err := io.ReadFull(r, buf)
+	if err != nil {
+		return n, err
+	}
+	return n, p.UnmarshalBinary(buf)
+}
+
+func (p *G1Elt) Equal(p2 kyber.Point) bool { x := p2.(*G1Elt); return p.inner.Equal(&x.inner) }
+
+func (p *G1Elt) Null() kyber.Point {
+	p.inner.X.SetZero()
+	p.inner.Y.SetOne()
+	p.inner.Z.SetZero()
+	return p
+}
+
+func (p *G1Elt) Base() kyber.Point {
+	p.inner, _, _, _ = bls12381.Generators()
+	return p
+}
+
+func (p *G1Elt) Pick(rand cipher.Stream) kyber.Point {
+	var buf [32]byte
+	rand.XORKeyStream(buf[:], buf[:])
+	return p.Hash(buf[:])
+}
+
+func (p *G1Elt) Set(p2 kyber.Point) kyber.Point { p.inner = p2.(*G1Elt).inner; return p }
+
+func (p *G1Elt) Clone() kyber.Point { return new(G1Elt).Set(p) }
+
+func (p *G1Elt) EmbedLen() int {
+	panic("bls12-381: unsupported operation")
+}
+
+func (p *G1Elt) Embed(_ []byte, _ cipher.Stream) kyber.Point {
+	panic("bls12-381: unsupported operation")
+}
+
+func (p *G1Elt) Data() ([]byte, error) {
+	panic("bls12-381: unsupported operation")
+}
+
+func (p *G1Elt) Add(a, b kyber.Point) kyber.Point {
+	aa, bb := a.(*G1Elt), b.(*G1Elt)
+	p.inner.Set(&aa.inner)
+	p.inner.AddAssign(&bb.inner)
+	return p
+}
+
+func (p *G1Elt) Sub(a, b kyber.Point) kyber.Point {
+	aa, bb := a.(*G1Elt), b.(*G1Elt)
+	p.inner.Set(&aa.inner)
+	p.inner.SubAssign(&bb.inner)
+	return p
+}
+
+func (p *G1Elt) Neg(a kyber.Point) kyber.Point {
+	p.inner.Neg(&a.(*G1Elt).inner)
```

Comment:
All of these are not safe for concurrent use since they aren't cloning `p` but setting it / overwriting it, right? 
So I guess this might warrant a comment / require to open an issue about it.

Either that or `p.inner` should be cloned prior to modification if this wasn't the intention.

Reply:
None of them are thread-safe.
I wouldn't expect any thread safety, and the bigint semantics (which this resembles) in Go are that the struct will be clobbered.

The straightforward reason for this is that it gives options to either clone before or reuse and save on allocations.
I can add a note.

Reply:
Long live rust, it would throw a tantrum if you use code like this in a threaded environment :) Last time I did some coding in go after a long and tedious rust session, i was like: "Why am I allowed to write this code? Why doesn't the compiler complain?"